torch/csrc/jit/codegen/cuda/kernel_cache.h

```diff
@@ -13,21 +13,125 @@ namespace jit {
 namespace fuser {
 namespace cuda {
 
+// Go through a tensor, and grab it's sizes/strides potentially broadcasted
+struct ExtractSizeStride {
+  std::vector<int64_t> sizes;
+  std::vector<int64_t> strides;
+
+  // TODO: broadcasted_size should be handled in codegen directly instead of at
+  //       the integration limbo.
+  explicit ExtractSizeStride(
+      const at::Tensor& val,
+      c10::optional<at::IntArrayRef> broadcasted_size = c10::nullopt) {
+    if (broadcasted_size) {
+      // [Note - broadcast support in integration]
+      // PyTorch follows numpy broadcasting rule.
+      // (https://numpy.org/doc/stable/user/basics.broadcasting.html)
+      //
+      // So in case where the rank of two operators differ, we align them on
+      // the higher dimensions, hence the offset o_dim-b_dim to the index here.
+      int b_dim = (int)broadcasted_size->size();
+      int o_dim = (int)val.dim();
+      TORCH_CHECK(b_dim >= o_dim);
+      for (int i = 0; i < b_dim; i++) {
+        sizes.push_back(broadcasted_size->at(i));
+        int index = i + o_dim - b_dim;
+        if (index < 0) {
+          strides.push_back(0);
+        } else if (val.sizes()[index] == sizes[i]) {
+          strides.push_back(val.strides()[index]);
+        } else {
+          TORCH_CHECK(
+              val.sizes()[index] == 1,
+              "Not compatible dimension size for broadcast");
+          strides.push_back(0);
+        }
+      }
+    } else {
+      auto o_dim = val.dim();
+      for (decltype(val.dim()) i{0}; i < o_dim; i++) {
+        sizes.push_back(val.sizes()[i]);
+        strides.push_back(val.strides()[i]);
+      }
+    }
+  }
+};
+
+class CudaKernel {
+ public:
+  void setFusionPtr(std::unique_ptr<Fusion> fusion) {
+    fusion_ = std::move(fusion);
+  }
+
+  Fusion* getFusionPtr() {
+    return fusion_.get();
+  }
+
+  const Fusion* getFusionPtr() const {
+    return fusion_.get();
+  }
+
+  CUmodule& getModule() {
+    return module_;
+  }
+
+  CUfunction& getFunction() {
+    return function_;
+  }
+
+  int16_t& getDevice() {
+    return device_;
+  }
+
+  const int16_t& getDevice() const {
+    return device_;
+  }
+
+  bool hasRNG() {
+    if (fusion_) {
+      FusionGuard fg(fusion_.get());
+      return fusion_->hasRNG();
+    }
+    return false;
+  }
+
+ private:
+  int16_t device_;
+  CUmodule module_;
+  CUfunction function_;
+  std::unique_ptr<Fusion> fusion_;
+
+  // WARNING:
+  // Block and Grid dimension setting is here for testing purposes only
+  // These are not here for general use and only for use with
+  // the runTestKernel() function.
+ public:
```

Comment:
1. public, then private then public makes it hard to read (ex. figure out the public interface)
  -> move the public functions to the public section above?

2. do block_ and grid_ also have to be public?

3. if block() and grid() are only used by runTestKernel() we may have better options:
a. private + friend runTestKernel() - easy option
b. create a testing interface (so we don't have to friend eveything in this class) - may not be worth it here

Reply:
1. block / grid is only used by testing. I'm hoping that they will go away in a few round of PRs as Kevin and I are merging code path. Hence the intention to leave them in the corner.
2. I think so, `block_` and `grid_` are used as accessors.
3. both option sounds reasonable.

Would you mind if I leave it as-is and do it in a following PR to merge `runKernel` and `runTestKernel`? Hence this wouldn't be needed any more.